File: app/src/test/java/io/intrepid/contest/screens/contestcreation/categorieslist/CategoriesListPresenterTest.java
package io.intrepid.contest.screens.contestcreation.categorieslist;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import java.util.ArrayList;
import java.util.List;

import io.intrepid.contest.models.Category;
import io.intrepid.contest.models.Contest;
import io.intrepid.contest.testutils.BasePresenterTest;

import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class CategoriesListPresenterTest extends BasePresenterTest<CategoriesListPresenter> {
    @Mock
    CategoriesListContract.ContestCreationFragment mockView;
    @Mock
    Contest.Builder mockContestBuilder;

    @Before
    public void setup() {
        when(mockContestBuilder.getCategories()).thenReturn(makeCategories());
        presenter = new CategoriesListPresenter(mockView,
                                                testConfiguration,
                                                mockContestBuilder);
    }

    private List<Category> makeCategories() {
        List<Category> categories = new ArrayList<>();
        for (int i = 0; i < 5; i++) {
            categories.add(new Category("Category " + i, "Category description " + i));
        }
        return categories;
    }

    @Test
    public void presenterShouldAddDefaultCategoryIfCategoriesIsEmptyOrNull() {
        when(mockContestBuilder.getCategories()).thenReturn(new ArrayList<>());
        presenter = new CategoriesListPresenter(mockView, testConfiguration, mockContestBuilder);
        verify(mockView).getDefaultCategory(anyInt(), anyInt());
    }

    @Test
    public void onViewBoundShouldEnableNextWhenCategoriesIsNotEmpty() {
        presenter.onViewBound();
        verify(mockView).setNextEnabled(true);
    }

    @Test
    public void onViewBoundShouldDisableNextWhenCategoriesIsEmpty() {
        when(mockContestBuilder.getCategories()).thenReturn(new ArrayList<>());
        presenter.onViewBound();
        verify(mockView).setNextEnabled(false);
    }

    @Test
    public void onDeleteClickedShouldNeverDisableWhenCategoriesIsNotEmpty() {
        List<Category> categories = mockContestBuilder.getCategories();
        for (int i = 0; i < categories.size() - 1; i++) {
            presenter.onDeleteClicked(categories.get(i));
        }

        verify(mockView, never()).setNextEnabled(false);
        verify(mockView, times(categories.size() - 1)).setNextEnabled(true);
    }

    @Test
    public void onDeleteClickedShouldDisableNextWhenCategoriesIsEmpty() {
        List<Category> singleCategoryList = new ArrayList<>();
        Category singleCategory = new Category("Single Category", "TEST");
        singleCategoryList.add(singleCategory);
        when(mockContestBuilder.getCategories()).thenReturn(new ArrayList<>());

        presenter.onDeleteClicked(singleCategory);

        verify(mockView).setNextEnabled(false);
        verify(mockView, never()).setNextEnabled(true);
    }

    @Test
    public void onViewShouldCreatedShouldTriggerViewToShowCategories() {
        presenter.onViewCreated();
        verify(mockView).showCategories(any());
    }

    @Test
    public void onViewBoundShouldTriggerViewToSetNextEnabled() {
        presenter.onViewBound();
        verify(mockView).setNextEnabled(true);
    }

    @Test
    public void onViewBoundShouldTriggerViewToDisableNextWhenCategoriesIsEmpty() {
        when(mockContestBuilder.getCategories()).thenReturn(new ArrayList<>());
        presenter.onViewBound();
        verify(mockView).setNextEnabled(false);
    }

    @Test
    public void displayCategoriesShouldShowDefaultCategoryWhenThereAreNoCategories() {
        ArrayList<Category> categories = new ArrayList<>();
        categories.add(new Category("Default name", "Default description"));
        when(mockContestBuilder.getCategories()).thenReturn(categories);

        presenter.displayCategories();

        verify(mockView).showCategories(argThat(argument -> argument.equals(categories)));
    }

    @Test
    public void displayCategoriesShouldShowCategoriesWhenThereAreCategories() {
        List<Category> categories = makeCategories();
        when(mockContestBuilder.getCategories()).thenReturn(categories);

        presenter.displayCategories();

        verify(mockView).showCategories(categories);
    }

    @Test
    public void onAddCategoryClickedShouldCauseViewToShowAdd() {
        presenter.onAddCategoryClicked();
        verify(mockView).showAddCategoryScreen();
    }

    @Test
    public void onNextClickedShouldTriggerViewToShowNextScreen() {
        presenter.onNextClicked();
        verify(mockView).showNextScreen();
    }

    @Test
    public void onCategoryClickedShouldTriggerViewToShowEditPage() {
        List<Category> categories = new ArrayList<>();
        categories.add(new Category("Test", "Tester"));
        when(mockContestBuilder.getCategories()).thenReturn(categories);

        presenter.onCategoryClicked(categories.get(0));

        verify(mockView).showEditCategoryPage(categories.get(0), 0);
    }

    @Test
    public void onDeleteClickedShouldTriggerViewToLoseCategory() {
        int initialSize = mockContestBuilder.getCategories().size();
        presenter.onDeleteClicked(mockContestBuilder.getCategories().get(0));
        verify(mockView).showCategories(argThat(argument -> argument.size() == initialSize - 1));
    }

    @Test
    public void onCategoryMovedShouldUpdateCategoriesListWhenItemMovedUp() {
        List<Category> categories = makeCategories();
        when(mockContestBuilder.getCategories()).thenReturn(categories);

        presenter.onCategoryMoved(1, 4);

        assertTrue(categories.get(4).getName().equals("Category 1"));
        assertTrue(categories.get(1).getName().equals("Category 2"));
    }

    @Test
    public void onCategoryMovedShouldUpdateCategoriesListWhenItemMovedDown() {
        List<Category> categories = makeCategories();
        when(mockContestBuilder.getCategories()).thenReturn(categories);

        presenter.onCategoryMoved(4, 1);

        assertTrue(categories.get(1).getName().equals("Category 4"));
        assertTrue(categories.get(4).getName().equals("Category 3"));
    }
}

File: app/src/test/java/io/intrepid/contest/screens/contestcreation/categorieslist/CategoriesListPresenterTest.kt
package io.intrepid.contest.screens.contestcreation.categorieslist

import io.intrepid.contest.models.Category
import io.intrepid.contest.models.Contest
import io.intrepid.contest.screens.contestcreation.categorieslist.CategoriesListContract.View
import io.intrepid.contest.testutils.BasePresenterTest
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.argThat
import org.mockito.Mock
import org.mockito.Mockito.never
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import java.util.*

class CategoriesListPresenterTest : BasePresenterTest<CategoriesListPresenter>() {
    @Mock
    private lateinit var mockView: View

    private lateinit var contestBuilder: Contest.Builder

    @Before
    fun setup() {
        contestBuilder = Contest.Builder().apply {
            categories = makeCategories()
        }
        presenter = CategoriesListPresenter(mockView, testConfiguration, contestBuilder)
    }

    @Test
    fun presenterShouldAddDefaultCategoryIfCategoriesIsEmptyOrNull() {
        contestBuilder.categories = mutableListOf()
        presenter = CategoriesListPresenter(mockView, testConfiguration, contestBuilder)
        verify<View>(mockView).getDefaultCategory(anyInt(), anyInt())
    }

    @Test
    fun onViewBoundShouldEnableNextWhenCategoriesIsNotEmpty() {
        presenter.onViewBound()
        verify<View>(mockView).setNextEnabled(true)
    }

    @Test
    fun onViewBoundShouldDisableNextWhenCategoriesIsEmpty() {
        contestBuilder.categories = mutableListOf()
        presenter.onViewBound()
        verify<View>(mockView).setNextEnabled(false)
    }

    @Test
    fun onDeleteClickedShouldNeverDisableWhenCategoriesIsNotEmpty() {
        val categories = contestBuilder.categories

        var i = 0
        while (i < (categories.size - 1)) {
            presenter.onDeleteClicked(categories[i++])
        }

        verify<View>(mockView, never()).setNextEnabled(false)
        verify<View>(mockView, times(categories.size - 1)).setNextEnabled(true)
    }

    @Test
    fun onDeleteClickedShouldDisableNextWhenCategoriesIsEmpty() {
        contestBuilder.categories = mutableListOf()
        val singleCategory = Category("Single Category", "TEST")

        presenter.onDeleteClicked(singleCategory)

        verify<View>(mockView).setNextEnabled(false)
        verify<View>(mockView, never()).setNextEnabled(true)
    }

    @Test
    fun onViewShouldCreatedShouldTriggerViewToShowCategories() {
        presenter.onViewCreated()
        verify<View>(mockView).showCategories(any<List<Category>>())
    }

    @Test
    fun onViewBoundShouldTriggerViewToSetNextEnabled() {
        presenter.onViewBound()
        verify<View>(mockView).setNextEnabled(true)
    }

    @Test
    fun onViewBoundShouldTriggerViewToDisableNextWhenCategoriesIsEmpty() {
        contestBuilder.categories = mutableListOf()
        presenter.onViewBound()
        verify<View>(mockView).setNextEnabled(false)
    }

    @Test
    fun onViewBoundShouldTriggerViewToFocus() {
        presenter.onViewBound()
        verify<View>(mockView).onFocus()
    }

    @Test
    fun displayCategoriesShouldShowDefaultCategoryWhenThereAreNoCategories() {
        val categories = listOf(Category("Default name", "Default description"))
        contestBuilder.categories = categories

        presenter.displayCategories()

        verify<View>(mockView).showCategories(argThat<List<Category>> { argument -> argument == categories })
    }

    @Test
    fun displayCategoriesShouldShowCategoriesWhenThereAreCategories() {
        presenter.displayCategories()
        verify<View>(mockView).showCategories(contestBuilder.categories)
    }

    @Test
    fun onAddCategoryClickedShouldCauseViewToShowAdd() {
        presenter.onAddCategoryClicked()
        verify<View>(mockView).showAddCategoryScreen()
    }

    @Test
    fun onNextClickedShouldTriggerViewToShowNextScreen() {
        presenter.onNextClicked()
        verify<View>(mockView).showNextScreen()
    }

    @Test
    fun onCategoryClickedShouldTriggerViewToShowEditPage() {
        val categories = listOf(Category("Test", "Tester"))
        contestBuilder.categories = categories

        presenter.onCategoryClicked(categories[0])

        verify<View>(mockView).showEditCategoryPage(categories[0], 0)
    }

    @Test
    fun onDeleteClickedShouldTriggerViewToLoseCategory() {
        val initialSize = contestBuilder.categories.size
        presenter.onDeleteClicked(contestBuilder.categories[0])
        verify<View>(mockView).showCategories(argThat<List<Category>> { argument -> argument.size == initialSize - 1 })
    }

    @Test
    fun onCategoryMovedShouldUpdateCategoriesListWhenItemMovedUp() {
        presenter.onCategoryMoved(1, 4)

        assertTrue(contestBuilder.categories[4].name == "Category 1")
        assertTrue(contestBuilder.categories[1].name == "Category 2")
    }

    @Test
    fun onCategoryMovedShouldUpdateCategoriesListWhenItemMovedDown() {
        presenter.onCategoryMoved(4, 1)

        assertTrue(contestBuilder.categories[1].name == "Category 4")
        assertTrue(contestBuilder.categories[4].name == "Category 3")
    }

    private fun makeCategories(): List<Category> {
        val categories = ArrayList<Category>()
        for (i in 0..4) {
            categories.add(Category("Category " + i, "Category description " + i))
        }
        return categories
    }
}
